Refactor patch - navigation and related fixes

kamilkrzyskow opened the pull request.

ultrabug commented on mkdocs_static_i18n/plugin.py:
```diff
@@ -78,7 +81,31 @@ def on_nav(self, nav, config, files):
+
+            # assure presence of homepage
+            if self.current_language == self.default_language:
+                i18n_nav.homepage = "/"
```
homepage should be a `Optional[Page]`, not `str` please

I also wonder about this, why would you set a default to the default language while you could also do the same for all of them?

I'd rather not set a default tbh.

kamilkrzyskow replied:
I printed the values of `item.url` and my brain 🥶. If it's supposed to be THE homepage, then I need to look at it again, and actually figure it out 😮‍💨.
I wanted to have a default value here, because it isn't being looked up currently and is `None` and there is the warning, which would break strict. 
My docs have the index page inside of a Section, which is generated by the awesome-pages plugin (the structure was "same" before the switch to using that plugin).

ultrabug replied:
We are looking for the homepage on the reconfigre L259 (of your code). I'm not sure what you're talking about?

kamilkrzyskow replied:
Yes it's being searched for on L259, but it's not being found for the default language `/` path in my `docs`.
I have the warning, which would break `strict` mode. 
I wanted to solve it easily, but the idea was wrong as you pointed out, investigating further...

kamilkrzyskow replied:
So after my investigation it turned out that:
```py
on_nav Page(title=[blank], url='/mkdocs-static-i18n/')
# ...
on_nav None
```
My docs don't have the homepage at the start of `on_nav`, possibly `awesome-pages` blocks it to then assign it during nav generation ;/
Don't know what to do at this moment to fix it, but I shall just commit my changes. It's either something to be done on `awesome-page`'s end or I could just fork it later and remove the warning.
Anyways...

ultrabug commented on mkdocs_static_i18n/plugin.py:
```diff
@@ -78,7 +81,31 @@ def on_nav(self, nav, config, files):
+                f"{'s' if NavHelper.translated_items > 1 else ''} to '{self.current_language}'"
+            )
+
+        if hasattr(i18n_nav, "homepage"):
```
since i18n_nav is a `Navigation`, shouldn't it have a `homepage` always? This if looks useless after your refactoring

kamilkrzyskow replied:
True

ultrabug commented on mkdocs_static_i18n/suffix.py:
```diff
@@ -154,9 +158,8 @@ def on_files(self, files, config):
-        else:
-            if self.is_default_language_build:
-                i18n_files.append(file)
+        elif self.is_default_language_build or file.src_uri.startswith("assets/"):
```
Can you explain why you're adding `or file.src_uri.startswith("assets/")` please? I must admit that I don't understand.

ultrabug replied:
Okay, funny to see all those possible use cases :brain:

ultrabug commented on mkdocs_static_i18n/suffix.py:
```diff
@@ -53,7 +54,10 @@ def reconfigure_file(
     else:
         file.dest_path = file_dest_path
     file.abs_dest_path = os.path.normpath(os.path.join(config.site_dir, file.dest_path))
-    file.name = Path(file_dest_path).stem
```
This one impacts the "unlocalization" of files. Since we're removing language suffix from them, it's important to align their name as well

Before your patch :
```
INFO     -  reconfigure File(src_uri='topic2/1.1.filename.fr.html', dest_uri='de/topic2/1.1.filename.html', name='1.1.filename', url='de/topic2/1.1.filename.html') from
```

After your patch :
```
INFO     -  reconfigure File(src_uri='topic2/1.1.filename.fr.html', dest_uri='de/topic2/1.1.filename.html', name='1.1.filename.fr', url='de/topic2/1.1.filename.html') from
```

kamilkrzyskow replied:
OK, I shall see what I can do. Before my changes `file.name` was `index` for almost every file. This lead the `navigation.index` Material feature to break. Maybe something else influenced it 🤔 

BTW I've noticed that it's a debug log and you displayed it as INFO. Is there a way to limit the log output to the plugin / exclude other plugins?
`log.debug(f"reconfigure {file} from locale {file_locale}")`

ultrabug replied:
> This lead the navigation.index Material feature to break.

Well then there must be something wrong in my own testing. My understanding of this feature is that it allows to NOT show a useless indentation of the "index" section in the sidebar. With the initial file.name cleanup, this is what happens indeed but if I remove it (your proposed code), it does.

> BTW I've noticed that it's a debug log and you displayed it as INFO. Is there a way to limit the log output to the plugin / exclude other plugins?

No, so I just shamelessly locally switched it to INFO :blush:

kamilkrzyskow replied:
What do you mean, by useless section. If the feature is turned on it merges the `index` page with the section "header". We have a lot of unmerged and merged sections on our page.
https://github.com/squidfunk/mkdocs-material/blob/master/src/partials/nav-item.html#L65-L66
It uses the `nav_item.is_index`, which looks at the `file.name` being `index` xD

Basically this is incorrect:
```py
reconfigure 
    File(
    src_uri='topic1/named_file.en.md', 
    dest_uri='de/topic1/named_file/index.html', 
    name='index', 
    url='de/topic1/named_file/') 
from locale en
```
because the name is taken from the destination, which in most cases is `index.html`.
The best I can do right now is `file.name = PurePath(file.name).stem`, but then this breaks another link
```py
reconfigure 
    File(
    src_uri='topic2/1.1.filename.html', 
    dest_uri='de/topic2/1.1.filename.html', 
    name='1.1', 
    url='de/topic2/1.1.filename.html') 
from locale en
```
also the locale is set to `en` while in actuality it doesn't have the `en` suffix. iirc before the refactor the plugin used `None` here 🤔

ultrabug replied:
> What do you mean, by useless section. If the feature is turned on it merges the index page with the section "header".

I mean exactly that "merge = remove useless", your term is better.

ultrabug replied:
> also the locale is set to en while in actuality it doesn't have the en suffix. iirc before the refactor the plugin used None here

Yes, but it was confusing since we're actually interested by the locale we're building

ultrabug replied:
> We have a lot of unmerged and merged sections on our page.

How can I "see" this and reproduce? Could you propose to add some files to the test folder structure to demonstrate this?

kamilkrzyskow replied:
> How can I "see" this and reproduce? Could you propose to add some files to the test folder structure to demonstrate this?

For some reason I get slightly delayed notifications, and the PR comments don't reload ad hoc... anyways
Imo there is no need for additional files. If a file is called `index.md` inside of a Section, then it should be merge with the header, if it's not called `index.md` then it should not be merged with the Section.

https://github.com/Gothic-Modding-Community/gmc/tree/dev/docs/zengin/scripts/classes <- none of the files are merged with the header and it behaves like on your current mkdocs page, the Section header isn't a URL

https://github.com/Gothic-Modding-Community/gmc/tree/dev/docs/zengin/worlds <- index merges with the header the rest are beneath

https://gothic-modding-community.github.io/gmc/zengin/ interactive example 😄 

Also here comes the kicker, if a page merges with the Section, the title is still taken from the Section and not the page. This seems to actually work, if I seen it correctly when the `file.name` was broken.

ultrabug replied:
OK I need to think about how to test that, reproduce properly via a test and then fix it for good

kamilkrzyskow replied:
If you can access the files list in the test. Compare the source path with the `file.name`. Only README and index should have the `index` `file.name`. Good luck :v: 
I'll try to think about how to fix configuration across builds tomorrow, and I really want to try to modify the JavaScript. Both MkDocs and Material have a static implementation of it since years. Material could introduce some changes in the coming future, but still it shouldn't be something needing a lot of maintenance.
This way only `i18n` should require preservation.

kamilkrzyskow replied:
Also after sleeping on the issue at hand I figured. Why is `file.name` being something else than `index` suddenly a problem. If the "unlocalization" requires the `file.name` to be `Path(dest_path).stem` use the `.stem` value there instead of the `file.name` 😵

kamilkrzyskow commented:
Hi @ultrabug,
I've cleaned up the PR from failed commits, finally fixed the homepage assignment and added a fix to properly show the ultimate time log. I did a force push, but kept a backup of the old branch just in case. [Here you can see a comparison between them](https://github.com/kamilkrzyskow/i18n/compare/refactor/config-pr-bkp...kamilkrzyskow:i18n:refactor/config)

The homepage assignment was fixed with allowing to check for `item.url == ""` because from my testing I noticed that the main/default homepage path does always have an empty url inside the object. The issues I've been facing turned out to not be related to any plugin. Having the index homepage inside a sub Section caused MkDocs to ignore it in automatic assignment.

https://github.com/ultrabug/mkdocs-static-i18n/blob/29e169e062ac8635b4ff6497e0d3824de5cbd708/mkdocs_static_i18n/reconfigure.py#L264

As of the time logging, it showed the finished time from a inner build, and the log message was added to the DuplicateFilter, which blocked every other build time logging. I basically flipped it, I block the inner ones and allow the last one. :v:

https://github.com/ultrabug/mkdocs-static-i18n/blob/29e169e062ac8635b4ff6497e0d3824de5cbd708/mkdocs_static_i18n/plugin.py#L171

I think there won't be any more additions to this PR, if I have anything I shall keep it for another PR.

kamilkrzyskow commented:
I spoke too soon 😞I still got the urge to clean up and extract a part of the navigation homepage code from the reconfigure function back into the `on_nav` event. I also considered adding an option to disable the `unsupported lunr.js warning` completely for a given language, but ultimately didn't go through with it, I'll get back to the idea another time.
Wasting so much GitHub Action time...

ultrabug commented:
@kamilkrzyskow thanks for your patience and continued work. This PR is going too far IMHO since we did not settle and cover the tests of the initial WIP PR.

I have nothing against trying to change the way we build and trying to get those search indexes lean. But this should be done once we have a clear and functional initial WIP PR.

Else we'll end up iterating over and over while nothing gets properly tested.

I'm gonna comment my latest commit on my initial WIP PR since I'm happy about what I just pushed.

ultrabug commented:
> I thought I could help you release suffix support faster, but it seems that won't happen soon. 

I agree and I'm not satisfied with how we move forward together and I feel responsible for it.
I won't go into arguing about this PR and code copy/pasting as it won't be helping to write yet another comment about that.

So let's try something else **if you agree**: are you on discord or somewhere where we could discuss faster than Github comments maybe?

If not, I'll answer here and propose something hopefully better so we can get this major rewrite done once and for all.

Thank you, we'll make it :)

kamilkrzyskow commented:
> I agree and I'm not satisfied with how we move forward together and I feel responsible for it.

No need to feel responsible, apart of rewriting the code written by me instead of merging it twice (a cherry-pick would be enough if the whole PR wasn't good), I have nothing against your workflow, I'm just that kind of person, which needs to be interested in something to do it, and I'm not interested in the whole refactor to fully cooperate with you, just unlucky circumstances. ✌️ 

Also yes, I'm a daily Discord user, but I'm not willing to invest my time into a fully tested and complete refactor, so I guess there is no need for faster communication.
Initially I wanted to just comment about things that could be improved, but then as I said in my [comment here](https://github.com/ultrabug/mkdocs-static-i18n/pull/216#issuecomment-1503704948) my goal was to make the refactor functional (suffix structure specifically) to allow me to introduce Material's social cards into my team's documentation. 
This PR made it somewhat functional restoring nav links working, removing warnings etc., but it turned out that I missed something about how the "unlocalization" of files works, hence you introduced more changes after the PR.
I didn't want to commit to a strict procedural testing process, so I jumped ship and went back to the pre-refactor version and introduced support for `site_description` override and did some "bigger" changes too to facilitate the support of the Social Cards plugin, the testing suite still somehow works 🤷

> I shall open up a PR for this tomorrow.

Well, I did the changes, but didn't test them with the `folder` structure, and I'm not willing to write tests for it either (especially when the _newer_ testing approach from the Sitemap PR isn't merged). Therefore, anticipating yet another PR rejection, I just abandoned the idea, or let's say I'm still procrastinating on it 😉 

Later, if I have the time, I will comment on some parts of the code that I see issues with, so just like I did before, giving out unsolicited advice 😏 

---

As a side-note, it turned out that the social cards plugin uses the `page.file.src` to generate the path to put the images and the meta head tag url to that image. This became a problem for the pages in another locale that weren't yet translated, hence they used the original default language source file. This re-usage of the source path leads to overwriting of the default language images by the other locale (which uses a different `site_description`). 
Maybe you'll come up with a way to handle such cases better than my decorator wrapper approach, but to me this rather unique case requires it.

https://github.com/kamilkrzyskow/i18n/commit/af9b1fc81fc7e615f1987e29e832963f2a067532

ultrabug commented:
Ok @kamilkrzyskow thanks for clarifying ; I'm off this week, will move on next one hopefully

jonnyforestGIS commented:
I @ultrabug a gentle ping about the status of this PR. 
Do you have plans or time to review this PR and merge?

Cheers

ultrabug commented:
@jonnyforestGIS yes I have many things going on at the same time so I'm pushing this PR in the backlog

Another reason is that we are working on the MkDocs 1.5 release which will change some internals so I'd rather make this PR target the latest MkDocs version instead.

I should spend again some time on this PR by the end of the month!

jonnyforestGIS commented:
Hi @ultrabug thank you for your quick feedback and thank you for your effort.
Cheers

ultrabug commented:
Let's move on, merging to @kamilkrzyskow gets credits ofc

I'll update the code on the other PR.
